### ISSUE TITLE
Multi-Contrast Multi-Sessions Feature Potential Discussions.

### ISSUE THREAD

**Issue by dyt811** [ivadomed/ivadomed#1009]
### Discussed in https://github.com/ivadomed/ivadomed/discussions/1003

<div type='discussions-op-text'>

<sup>Originally posted by **dyt811** December 13, 2021</sup>
Hi all. 

We are finally at a sufficiently refactored and base code improved stage to merge earlier radical experimental branch of changes. This would be a good time to revisit some key assumptions which are needed to resolve challenges with adopting multi-sessions/contrasts support: 

## Key Assumptions: 

* All modality/session input data are coregistered by user already (and of same dimension!). 

* For a subject to be included, it must have ALL [`Targeted Modality`] x [`Targeted Session`] data. Missing any session or modality will result in exclusion. 
   * Rationale: if we start accommodating either missing modalities or missing sessions, the input vector will not be stable (since they need to be FIXED dimension) and will lead to require some very non-scalable solution to guess which session/modality to use for the missing ones or interpolation across modality/session which require many experiment specific knowledge. 
   * For the initial release, we focus on the bare minimal cases: all modality all session data are present, fixed and of **same dimension** at least within each subject. 
  
## Clarification Questions: 

* Currently, it seems we don't require all `target_suffix` data to be present (I believe?). Keeping this? 
   * Rationale: if all are data are coregistered with each other, at least ONE of the `target_suffix` (aka ground truth) is sufficient?  

* What about `roi_suffix`? Right now it is incorporated into `target_suffix` and the code just check if it is one of the derivatives detected? 

* Should we consider ground truth outside of the targeted modality? See below: Currently STRICT [target modality] match.
![image](https://user-images.githubusercontent.com/6720068/145754667-52378b70-9c00-48b3-b804-27fe02890947.png)
</div>

**Comment by dyt811** [comment]
Per answers gathered from the meeting: 

1) We want to keep this. As long as at least one ground truth is defined. We keep the subject. 
   * Example: if subject has t1 target_suffix when config parameters requires both t1 and t2 target suffix (a list of t1/t2). 

2) ROI suffix is used to define ROI region (as a kind of spatial filtering). Combining ROI suffix and Target suffix for indexing purpose is FINE BUT we cannot leave ROI suffix inside target suffix for ground truth definition. 
   * [The `suffix` property in `roi_params`](https://ivadomed.org/configuration_file.html?highlight=roi_params#roi-params) is not required for all studies 
   * `target_suffix` are essentially ground-truth, see [Documentation](https://ivadomed.org/configuration_file.html?highlight=target_suffix#target-suffix)
   * The `suffix` property in `roi_params` is not related to ground truth (for training purposes). It helps refines the ground truth to reduce region used for training. 
   * In the current code, `roi_suffix` is added to `target_suffix` for file indexation in bids_dataframe, but it **could** be kept separated if that is an issue


3) Yes, we should consider additional ground truth defined in target modality. 
   
   * ![image](https://user-images.githubusercontent.com/6720068/145876621-f1dd6997-610d-4c1e-85e5-9c69a28b3e5e.png) 
    Regarding the screenshot above: i.e. we will use the red bound boxes data. T2star ground truth. 
   * eg: modality T1w and T2w, if a GT only exists in FLAIR, we keep it).

**Comment by mariehbourget** [comment]
Hi @dyt811!
I re-read the questions and answers you gathered in this issue and everything look good to me and conform to what we discussed in meeting.

> ROI suffix is used to define ROI region (as a kind of spatial filtering). Combining ROI suffix and Target suffix for indexing purpose is FINE BUT we cannot leave ROI suffix inside target suffix for ground truth definition.
> 
> * [The `suffix` property in `roi_params`](https://ivadomed.org/configuration_file.html?highlight=roi_params#roi-params) is not required for all studies
> * `target_suffix` are essentially ground-truth, see [Documentation](https://ivadomed.org/configuration_file.html?highlight=target_suffix#target-suffix)
> * The `suffix` property in `roi_params` is not related to ground truth (for training purposes). It helps refines the ground truth to reduce region used for training.
> * In the current code, `roi_suffix` is added to `target_suffix` for file indexation in bids_dataframe, but it **could** be kept separated if that is an issue

For the last point, the reason why we added `roi_suffix` in `target_suffix` [here](https://github.com/ivadomed/ivadomed/blob/d6fe24db4d6a4319c57802b7a6d68e17f6d8b9e0/ivadomed/loader/bids_dataframe.py#L49) is only to do [this filtering operation](https://github.com/ivadomed/ivadomed/blob/d6fe24db4d6a4319c57802b7a6d68e17f6d8b9e0/ivadomed/loader/bids_dataframe.py#L125) which keeps derivatives that contain `target_suffix` or `roi_suffix`, and discard the others. As you mentioned, we could keep them separated and update the filtering accordingly if it's an issue with your developments.

**Comment by dyt811** [comment]
## Current PR 

First version of the PR will target the most common simple cases: 
* Multiple Sessions Data
* Multiple Modality Data
* Combination of the above
* Assume all subject specific volumes are co-reregistered
* No/limited file substitution support. 

Additional Key Assumptions: 
* Most have been implemented at the indexer level
* Most substitution are not implemented at the file pairing level (EVEN though it might have bene properly indexed)

1. Single ground truth labeling
   - Similarity Score based matching between closest possible ground truth and source data
   - We always assume that for a single input or single group of input there is at least one or more ground truth. 
   - Potentially substitutable across raters if at least one of the ground truth are provided. 
1. Session x Modality determines input
   - The configuration JSON specifies session size (assume 1 if not list) and modality list size (cannot be empty) determines the number input volumes to the neural network. 
   - Minimum amount of input volumes (session x modality) must be met. It is okay to have more sessions but cannot have less.
1. Missing session = disqualify
   - Session cannot substitute each other
   - Session CAN be left blank (indicating no session preferences) 
1. Missing modality = disqualify
   - Modality cannot substitute each other
   - Modality/Contrast CANNOT be left blank (must be specified else raise exception)
1. Missing ground truth = potentially can be substituded. 
   - Ground truth can substitute each other
   - Ground truth from all sessions (even unmentioned ones) are indexed

## Things to discuss: Most likely to be implemented in Phase II PRs.

* How/when/if to substitute **sessions**:
   * Do we permit across session same modality substitution
   * Do we permit across session cross modality substitution
   * How do we specify across session preferences? e.g. Expected input Ses [1, 2, 3] x [T1,T2], but only has Ses1-T1, Ses2-T1, Ses2-T2, Do we discard or keep? 
   * What about extra session information? e.g. Expected input Ses [2, 3] x [T1,T2], subject missing Ses2T1, but has Ses4-T1 and Ses1-T1, which one do we use to substitute?
* How/when/if to substitute **modalities**:
   * Due to input level dropout we MAY choose to retain missing modality subjects. 
   * However keep in mind that in multisession scenarios with multi modality dropout, we may be missing significant portion of data. 
   * Should we set a hard coded 50% minimum data input threshold. e.g. Ses [1, 2, 3] x [T1,T2] require at least 3 dataset to be included. 
*

**Comment by dyt811** [comment]
Update on this discussion. 

Per 2022-07-04 proposal at the ivadomed developer meeting and subsequent feedback sessions from @naga-karthik  and @mariehbourget, 

We believe to maximize flexibility, some significant reworks are required. 

Here is a list of compiled ideas, more details will be presented and updated here with more concrete example given. 

- Enable Dataset List[dict] entry under loader_parameters to support multiple datasets: 
- Datsets: [dict1, dict2, dict3]
- Within each of the dict have its own `label` key, and one of the following key
  - `training` 
  - `validation`
  - `testing`
  - `training_validation` (require split specified)
  - `training_validation_testing` (require split specified)
  - This allows people to manually separate train/val/test sources OR use a bucket and do a split from the provided source
- Each of the above tag, is a list of dict
- Dict type: BIDS: 
   - Leverage N number of BIDS keywords to filter for results, validated against expected N for ground truth and images, at path specified. 
- Dict type: REGEX:
   - Carry out ReGex operations, validated against expected N for ground truth and images, at path specified. 
- Dict type: FILES:
   - Hard coded lists.

**Comment by lifetheater57** [comment]
Based on my experience in a previous project, another way of loading samples would be to use `recipe`s to specify how to generate the training examples from the dataframe of the examples loaded. It could be used after loading either BIDS datasets or a user-provided dataframe with required information. 
 
Here is a definition of a `recipe`. Some instructions might be redundant with the current `ivadomed` config entries.
```yaml
recipe : dict {
    filter : Optional(dict) = entries with the `key: value` form where the key is the column to filter with and the value is the value that has to be matched (either matching a single value or a value out of a list of values) 
    combine : dict {
        function : str = function name (e.g. concat, add, prod)
        column : str = the column used for the selection of samples for combination
        sources : Union(list, str, recipe) = the value in the selected column used to be combined with the function (can be recipe to further prepare the samples for combination)
        GT : str = the index of the source whose GT will be used, random if None (e.g. 0, 2, None)
        match : Union(list, str) = the columns to use for the matching of corresponding entries when combining; Note: a recipe within another would use the union of its `match` entry and the `match` entry of the parent excluding the column specified in its `column` entry 
        imputation : Union(dict, str) = the method to use for imputing the missing values (e.g. drop, mean, zero, "some value") that can be specified for each column used either for matching or combining using a dictionary or globally using a string
    }
}
```
Below is an example of configuration in the above `recipe` format where we use data of 3 subjects, the `T1w`, `T2W` and the sum of sagittal and axial orientations of `T1w` are concatenated. To match the samples for the concatenation, the subject, orientation and session values are used. Samples are dropped when an entry (`subject`, `orientation`, `session`, `contrast`) is missing.
```json
"filter": {
    "subject": [
        "sapienza05", 
        "strasbourg04", 
        "unf01"
    ]
}
"combine": {
    "function": "concat",
    "column": "contrast",
    "sources": [
        "T1w", 
        "T2w",
        {
            "T1w" : {
                "combine": {
                    "function": "add",
                    "column": "orientation",
                    "sources": ["sag", "ax"],
                    "GT": "add",
                    "match": [],
                    "imputation": "drop"
                }
            }
        }
    ],
    "GT": null,
    "match": [
        "subject",
        "orientation",
        "session"
    ],
    "imputation": "drop"
}
```
Note that in this configuration, since orientation is not specified for the first "T1w" and the "T2w" sources then, assuming that both sagittal and axial images are availble, for a given subject and session, 2 samples would be generated:
- [T1w_sag, T2W_sag, T1w_sag+T1w_ax]
- [T1w_ax, T2W_ax, T1w_sag+T1w_ax]
